miio/curtain_babai.py

```diff
@@ -0,0 +1,201 @@
+        return self.motor_control == MotorControl.Open
+
+    @property
+    def is_close(self) -> bool:
+        """Device status."""
+        return self.motor_control == MotorControl.Close
+
+    @property
+    def mode(self) -> Mode:
+        """Motor rotation polarity."""
+        return Mode(self.data["mode"])
+
+    @property
+    def current_position(self) -> int:
+        """Run time of the motor."""
+        return self.data["current_position"]
+
+    @property
+    def target_position(self) -> int:
+        """Target curtain position."""
+        return self.data["target_position"]
+
+
+class CurtainBabai(MiotDevice):
+    """Main class representing the babai.curtain.190812 curtain."""
+
+    mapping = _MAPPING
+
+    @command(
+        default_output=format_output(
+            "",
+            "Motor control: {result.motor_control}\n"
+            "Motor mode: {result.mode}\n"
+            "Current position: {result.current_position}\n"
+            "Target position: {result.target_position}\n",
+        )
+    )
+    def status(self) -> CurtainStatus:
+        """Retrieve properties."""
+        # if use get_properties_for_mapping and max_len>1 then code -4004.
+        data = []
+        for k, v in self.mapping.items():
+            if "aiid" not in v:
+                data.extend(self.get_property(k))
+                # data[k] = prop["value"] if prop["code"] == 0 else None
+        return CurtainStatus(self._extract_property_value(data))
+
+    def get_property(self, name: str) -> List[Dict[str, Any]]:
+        if name not in self.mapping:
+            raise ValueError("key %s not in MAPING", name)
+        v = self.mapping[name]
+        try:
+            prop = self.get_property_by(siid=v["siid"], piid=v["piid"])
+        except DeviceError as e:
+            if e.code != -9999:
+                raise
+            sleep(5)
+            prop = self.get_property_by(siid=v["siid"], piid=v["piid"])
+        for p in prop:
+            p["did"] = name
+        return prop
+
+    @staticmethod
+    def _extract_property_value(prop: List[Dict[str, Any]]) -> Dict[str, Any]:
+        return {p["did"]: p["value"] if p["code"] == 0 else None for p in prop}
+
+    def _get_property_value(self, name: str) -> Any:
+        prop = self.get_property(name)
+        return self._extract_property_value(prop).get(name)
+
+    @property
+    def motor_control(self) -> MotorControl:
+        """Device status."""
+        return MotorControl(self._get_property_value("motor_control"))
+
+    @property
+    def mode(self) -> Mode:
+        """Motor rotation polarity."""
+        return Mode(self._get_property_value("mode"))
+
+    @property
+    def current_position(self) -> int:
+        """Run time of the motor."""
+        return self._get_property_value("current_position")
+
+    @property
+    def target_position(self) -> int:
+        """Target curtain position."""
+        return self._get_property_value("target_position")
```

Comment:
Properties should do no I/O, but I'd simply remove these as the information is available inside the status container.

Reply:
random code = -9999 from status vs 1 request